### PR TITLE
User history report UX updates

### PR DESCRIPTION
**Pull request by orangejenny**
## Summary
https://dimagi-dev.atlassian.net/browse/USH-1170

This makes most of the changes in the "Updated UI Changes 7/9" section of [the spec](https://docs.google.com/document/d/1Z5ZOu7O-BVOh62oCTc1JjNRsBAiMzs9i8Cf48q_9BDQ/edit#):

- Updates labels for filters and column headers
- Fixes issue where the reports navigation was being shown even though the page is in the users section
- Adds a filter by property
- Updates the changes display so a few main properties are shown on page load and others are a click away. I took some liberties with the spec here and expect this to change more, but the new look is an improvement over showing raw JSON.

Screen Shot 2021-07-15 at 12 08 28 PM

## Feature Flag
View user history report under user management

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

None

### QA Plan

Not requesting QA

### Safety story
These are all surface-level changes to a report that's currently flagged and not used by anyone external.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/reports/standard/users/reports.py
Comment (snopoke) [inline corehq/apps/reports/standard/users/reports.py:132]: this field isn't indexed, is that going to be a performance concern?

Comment (snopoke) [reply corehq/apps/reports/standard/users/reports.py:132]: Also, might be worth making a note in the model class about the expected structure of the JSON field.

File: corehq/apps/reports/filters/users.py
Comment (mkangia) [inline corehq/apps/reports/filters/users.py]: Can we avoid this branching? and keep "Profile" separate?

File: corehq/apps/reports/filters/users.py
Comment (mkangia) [inline corehq/apps/reports/filters/users.py]: this is primary location for just the mobile worker yes?

`location_id` and `assigned_location_ids` is only set for CommCareUser and not for WebUser since for web user these are stored on Domain Membership.
we should keep that clear here.
"Mobile User Primary Location" or "Mobile User Location" (since mobile user only has one location so 'Primary' isn't really needed)

File: corehq/apps/reports/filters/users.py
Comment (mkangia) [inline corehq/apps/reports/filters/users.py:479]: nit: can this be `user_property`? just to be more explicit. Even though the filter is for users but that is a bit more implicit

File: corehq/apps/reports/standard/users/reports.py
Comment (mkangia) [inline corehq/apps/reports/standard/users/reports.py:29]: nit: `get_user_history_primary_properties` since this for use for just the User History Report?

File: corehq/apps/reports/standard/users/reports.py
Comment (mkangia) [inline corehq/apps/reports/standard/users/reports.py:199]: Is it possible to put this in a html template and render that with context instead?
This method has grown to be quite complex and then is handling html render as well.

**Review by mkangia — COMMENTED** [review]
Looks great. Have couple of suggestions, queries, name changes.

File: corehq/apps/reports/standard/users/reports.py
Comment (snopoke) [inline corehq/apps/reports/standard/users/reports.py:177]: should this also use `format_html`?

File: corehq/apps/reports/standard/users/reports.py
Comment (snopoke) [inline corehq/apps/reports/standard/users/reports.py:180]: also here re using `format_html`

File: corehq/apps/reports/standard/users/reports.py
Comment (snopoke) [inline corehq/apps/reports/standard/users/reports.py:193]: more maybe usage of `format_html`

File: corehq/apps/reports/filters/users.py
Comment (orangejenny) [reply corehq/apps/reports/filters/users.py:470]: Good eye for detail. Added 40b650d7fc8889ed20c3eeaac68ca46e842b480e to add `(s)` since that's more accurate.

File: corehq/apps/reports/static/reports/js/user_history.js
Comment (orangejenny) [reply corehq/apps/reports/static/reports/js/user_history.js:7]: This check is a pattern that's used in other report-specific js, like `inspect_data.js`. I'm not worried about it changing if the URL changes, I think we'd grep for it and if not, it would be apparent that it's broken pretty quickly.

It's true that this class isn't necessarily specific to this report, so it could go in a more generic file. But to avoid bloat, my general principle is to put code in the most specific place possible, and then move it when other code starts using it.

File: corehq/apps/reports/static/reports/js/user_history.js
Comment (orangejenny) [reply corehq/apps/reports/static/reports/js/user_history.js:10]: c720d35e8bc

I wrote up a self-justifying comment and then decided I should just defer to your fresh eyes reading this code.

File: corehq/apps/reports/standard/users/reports.py
Comment (orangejenny) [reply corehq/apps/reports/standard/users/reports.py:168]: I'm going to say the current code is more accurate/specific, given that the class in question is called "unstyled".

File: corehq/apps/reports/filters/users.py
Comment (orangejenny) [reply corehq/apps/reports/filters/users.py]: I don't love combining these either, but looking at `details['changes']` doesn't tell me whether it was the profile or other user data that changed, and I don't want the filter to be parsing the change message text, as discussed in https://github.com/dimagi/commcare-hq/pull/30064#issuecomment-881481494

File: corehq/apps/reports/filters/users.py
Comment (orangejenny) [reply corehq/apps/reports/filters/users.py]: sure - c18483cf2df

File: corehq/apps/reports/standard/users/reports.py
Comment (orangejenny) [reply corehq/apps/reports/standard/users/reports.py]: oops, I lost track of where this function was, I'd thought it was in a users file. Updated in 62a20d3fbfd to make it a class method of the report, which namespaces it.

File: corehq/apps/reports/standard/users/reports.py
Comment (orangejenny) [reply corehq/apps/reports/standard/users/reports.py]: 4cbabbf6e58

File: corehq/apps/reports/standard/users/reports.py
Comment (orangejenny) [reply corehq/apps/reports/standard/users/reports.py:176]: It can't because `value` might be HTML, but it should be all right because `value` goes through `format_html` in the lines above.

File: corehq/apps/reports/standard/users/reports.py
Comment (orangejenny) [reply corehq/apps/reports/standard/users/reports.py]: 74af9399b2c - I don't really know if it'll be a performance concern, I'm just kind of assuming that this table will get quite large and that searching a nested JSON field will be slow. If you have any general thoughts on when to add indices, I'm interested.

File: corehq/apps/reports/filters/users.py
Comment (orangejenny) [reply corehq/apps/reports/filters/users.py]: Updated in b1ab22c5104 to specify this is just the primary location and only for mobile users.

Is there a plan for adding domain membership changes? I can't remember if we discussed this already, but looking at it now, it seems odd that web user location changes show up in the message but don't have anything in the changes dict.

Screen Shot 2021-07-16 at 8 32 58 PM

File: corehq/apps/reports/standard/users/reports.py
Comment (mkangia) [reply corehq/apps/reports/standard/users/reports.py]: can you confirm if that new index is actually used when we run this query?

File: corehq/apps/reports/standard/users/reports.py
Comment (mkangia) [reply corehq/apps/reports/standard/users/reports.py]: thank you :)

File: corehq/apps/reports/filters/users.py
Comment (mkangia) [reply corehq/apps/reports/filters/users.py]: > Is there a plan for adding domain membership changes?

No. No plans to include changes in related models as direct property changes.

> I can't remember if we discussed this already, but looking at it now, it seems odd that web user location changes show up in the message but don't have anything in the changes dict.

Currently the report is showing exactly what is in the model, and the model keeps "related model's change messages" separately. The UI definitely can show it however it's better for UI/UX.

File: corehq/apps/reports/filters/users.py
Comment (mkangia) [reply corehq/apps/reports/filters/users.py]: Actually, taking a step back.
My understanding was that "profile" was a part of the "user_data" itself. Is "profile" as special user data?
I think I am confused by the need to mention profile separately.

File: corehq/apps/reports/standard/users/reports.py
Comment (mkangia) [reply corehq/apps/reports/standard/users/reports.py]: being addressed in https://github.com/dimagi/commcare-hq/pull/30058#discussion_r671046694

File: corehq/apps/reports/standard/users/reports.py
Comment (mkangia) [reply corehq/apps/reports/standard/users/reports.py:167]: sounds reasonable.

I was just thinking of it as adding proper tags to make it "pretty" in html.
Its just personal preference to avoid negative bool (if that is the right way to refer this)

File: corehq/apps/reports/static/reports/js/user_history.js
Comment (mkangia) [reply corehq/apps/reports/static/reports/js/user_history.js]: thank you :)

File: corehq/apps/reports/static/reports/js/user_history.js
Comment (mkangia) [reply corehq/apps/reports/static/reports/js/user_history.js:7]: > It's true that this class isn't necessarily specific to this report, so it could go in a more generic file. But to avoid bloat, my general principle is to put code in the most specific place possible, and then move it when other code starts using it.

Sorry, i just meant that if the reason for adding a url check was to avoid conflicting with DOM elements with same class/Id on a different page, we could just keep the class/id naming very specific and then not need to add URL check at all.

File: corehq/apps/reports/standard/users/reports.py
Comment (snopoke) [reply corehq/apps/reports/standard/users/reports.py]: Also, I think it would be better to use `has_key`: https://docs.djangoproject.com/en/2.2/ref/contrib/postgres/fields/#std:fieldlookup-hstorefield.has_key

```
Q(details__changes__has_key=property)
```

(Would just need to confirm that it works for keys that are nested)

File: corehq/apps/reports/static/reports/js/user_history.js
Comment (orangejenny) [reply corehq/apps/reports/static/reports/js/user_history.js:7]: Going to leave this as-is - I know it would work either way, but the URL checking is a pattern I want to reinforce.

File: corehq/apps/reports/filters/users.py
Comment (orangejenny) [reply corehq/apps/reports/filters/users.py]: You mean just remove the "Profile or" part? I think that projects that use profiles may be looking for a "profile" option.  But I can ask product about it (don't want to block this PR on that, though).

I'm guessing we'll iterate more on the right way to display profile and user data changes. "User Data" also isn't great: we don't use that text consistently in the UI, only in bulk upload.

File: corehq/apps/reports/filters/users.py
Comment (orangejenny) [reply corehq/apps/reports/filters/users.py]: Hmm, I think it'll be confusing for users that they can only see locations for mobile users, not web users. But we don't need to deal with that right now.

File: corehq/apps/reports/standard/users/reports.py
Comment (orangejenny) [reply corehq/apps/reports/standard/users/reports.py]: Updated to use `has_key` in 88295f557a4 and verified it works.

> can you confirm if that new index is actually used when we run this query?

Good call. It is not used. Docs suggest a GIN index would be the way to best index this field, but it looks like it would need a separate index for each individual field. That's 8 new indices that would get updated every time a user is saved. Given that bulk user uploads are already a performance issue and that this report is new and we don't know which filtering will be most valuable, I'm inclined to not add any indexes at this point. ed70e2ff984 removes the unhelpful B-tree index I had added. Open to feedback.

File: corehq/apps/reports/filters/users.py
Comment (mkangia) [reply corehq/apps/reports/filters/users.py]: ya, my assumption is that profile is simply a `user_data`. The requirement that it needs to be treated as a special one, isn't great. If we can push back on it, that would be great.

> (don't want to block this PR on that, though).

Cool.

Comment (mkangia) [reply corehq/apps/reports/filters/users.py]: > I'm guessing we'll iterate more on the right way to display profile and user data changes. "User Data" also isn't great: we don't use that text consistently in the UI, only in bulk upload.

Ya. Iteration can happen on how to show these on UI but I hope we won't need to store it differently.

File: corehq/apps/reports/filters/users.py
Comment (mkangia) [reply corehq/apps/reports/filters/users.py]: searching would be limited to mobile workers location but users should be able to see locations on both mobile and web in the report messages.

**Comment by orangejenny** [comment]
Ah, I missed that. I might argue to product that we should remove location
from this filter, then, as it might become even more confusing that you can
see location but can't filter by it.

File: corehq/apps/reports/standard/users/reports.py
Comment (mkangia) [reply corehq/apps/reports/standard/users/reports.py]: +1 to reverting the index and see how much this hits performance.

confession, Searching by each property was not something I expected to come on this report, hence it was not structured to be searchable.

Comment (snopoke) [reply corehq/apps/reports/standard/users/reports.py]: You would only need a single GIN index which would cover the whole field: https://www.postgresql.org/docs/current/datatype-json.html#JSON-INDEXING

It shouldn't be hard to create one manually in a Django migration.

Comment (orangejenny) [reply corehq/apps/reports/standard/users/reports.py]: Hmm, so to experiment with this I created this index locally in `dbshell`:

`create index users_userh_details_idx on users_userhistory using gin (details);`

But when querying, it doesn't seem to get used. These queries all still do sequential scans:

```
In [12]: UserHistory.objects.filter(details__isnull=False).values("details").explain()
Out[12]: 'Seq Scan on users_userhistory  (cost=0.00..1.06 rows=6 width=32)\n  Filter: (details IS NOT NULL)'

In [13]: UserHistory.objects.filter(details__changes__isnull=False).values("details").explain()
Out[13]: "Seq Scan on users_userhistory  (cost=0.00..1.07 rows=6 width=32)\n  Filter: ((details -> 'changes'::text) IS NOT NULL)"

In [14]: UserHistory.objects.filter(details__changes__contains="a").values("details").explain()
Out[14]: 'Seq Scan on users_userhistory  (cost=0.00..1.09 rows=1 width=32)\n  Filter: ((details -> \'changes\'::text) @> \'"a"\'::jsonb)'
```

I can look more tomorrow, but curious if I"m doing something obviously wrong.

Comment (snopoke) [reply corehq/apps/reports/standard/users/reports.py]: The Django queries aren't using the supported index operations:

> The default GIN operator class for jsonb supports queries with top-level key-exists operators ?, ?& and ?| operators and path/value-exists operator @>.

From the Django docs the query functions that are available are:
* contains (`@>`)
  * you did test this but I think you'd need to pass a dict instead of just a key, also I tested this below and it doesn't work for nested objects. https://www.postgresql.org/docs/current/datatype-json.html#JSON-CONTAINMENT
* has_key (`?`)
* has_keys (`?&`)
* has_any_keys (`?|`)

I think the only question is whether or not the index can be used for nested keys. The docs seems to say no and I've confirmed that with a quick shell test:
```sql
-- we can find 'c' because it's a top-level key
select '{"c": {"d": 3}}'::jsonb ? 'c'; -- yields true

-- we can't find 'd' because it's not a top-level key
select '{"c": {"d": 3}}'::jsonb ? 'd'; -- yields false

-- we also can't check for containment since it doesn't support rested objects (and we'd need to know the value)
SELECT '{"foo": {"bar": "baz"}}'::jsonb @> '{"bar": "baz"}'::jsonb;  -- yields false
```

So with the current structure of `details` there is no way to efficiently do the query you are trying to do.

Comment (snopoke) [reply corehq/apps/reports/standard/users/reports.py]: Made some notes in the design doc (https://docs.google.com/document/d/15WS2OuB8l8UJePuwnZASIerxIrB2Gfnf1RRt24O3n8M/edit?disco=AAAANIegoeI) but just noting here as well that it may make sense to put the list of changed fields in a separate DB field for the purpose of efficient querying.

Comment (orangejenny) [reply corehq/apps/reports/standard/users/reports.py]: That makes sense, thank you for the research.

**Comment by orangejenny** [comment]
@snopoke @mkangia I think this is good to merge now.

File: corehq/apps/users/migrations/0031_index_userhistory_details.py
Comment (mkangia) [inline corehq/apps/users/migrations/0031_index_userhistory_details.py:15]: are we still keeping the index? I thought this was not getting used on searching.

**Review by mkangia — APPROVED** [review]
looks good. Just a confirmation on the index.

File: corehq/apps/users/migrations/0031_index_userhistory_details.py
Comment (orangejenny) [reply corehq/apps/users/migrations/0031_index_userhistory_details.py]: oops, forgot to push changes, thanks

**Comment by mkangia** [comment]
Thank you for making this report so much better :100:

**Pull request merged**
